src/Simulation/Simulators/SparseSimulator/Native/factory.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Manages simulators in a vector of pointers to simulators
+
+#include "factory.hpp"
+#include "SparseSimulator.h"
+#include "types.h"
+#include <iostream>
+#include <shared_mutex>
+
+
+namespace Microsoft
+{
+namespace Quantum
+{
+namespace SPARSESIMULATOR
+{
+// Ensures exclusive access to _simulators, the vector of simulators
+std::shared_mutex _mutex;
+std::vector<std::shared_ptr<SparseSimulator>> _simulators;
+
+unsigned createSimulator(logical_qubit_id num_qubits)
+{
+    if (num_qubits > MAX_QUBITS)
+        throw std::runtime_error("Max number of qubits is 1024!");
+    std::lock_guard<std::shared_mutex> lock(_mutex);
+    size_t emptySlot = -1;
+    for (auto const& s : _simulators)
+    {
+        if (s == NULL)
+        {
+            emptySlot = &s - &_simulators[0];
+            break;
+        }
+    }
+    if (emptySlot == -1)
+    {
+        _simulators.push_back(std::shared_ptr<SparseSimulator>(new SparseSimulator(num_qubits)));
+        emptySlot = _simulators.size() - 1;
+    }
+    else
+    {
+        _simulators[emptySlot] = std::shared_ptr<SparseSimulator>(new SparseSimulator(num_qubits));
```

Comment:
FYI: [Item 21: Prefer `std::make_unique` and `std::make_shared` to direct use of `new`](https://moodle.ufsc.br/pluginfile.php/2377667/mod_resource/content/0/Effective_Modern_C__.pdf).